### ISSUE TITLE
override a method with lesser arguments don't generate a error

### ISSUE THREAD

**Issue by jannesen** [Microsoft/TypeScript#26533]
**TypeScript Version:**  3.0.1 with --strict

```ts
class A {
    public m(a:string|null): string {
        return "";
    }
}

class B extends A {
    public m(a:string): string {
        return "";
    }
}
```
**Expected behavior:**
Error message

**Actual behavior:**
No error

B.m override the method A.m so the following code is valid.

```ts
let a:A = new B()
a.m(null);
```

But now B.m is called with null this violates the typing.

**Comment by jannesen** [comment]
according to https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html with --strictFunctionTypes parameter type checking is contravariantly. So subtyping a argument is not allowed anymore.

I use the --strict compiler option

**Comment by mattmccutchen** [comment]
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html says "The stricter checking applies to all function types, except those originating in method or constructor declarations."

**Comment by kitsonk** [comment]
IIRC that was because the in the wild code that it broke was rather extensive and because we have bad code, this is why we can't have nice things.

**Comment by RyanCavanaugh** [comment]
> this is why we can't have nice things

Indeed; the DOM API itself is unsound in this regard unless you're willing to entertain the notion of write-only properties.

**Comment by jannesen** [comment]
With the response from mattmccutchen I was able find the following issues [18654](https://github.com/Microsoft/TypeScript/pull/18654), [22156](https://github.com/Microsoft/TypeScript/issues/22156), [18963](https://github.com/Microsoft/TypeScript/issues/18963#issuecomment-334586832)

In #18654 @ahejlsberg wrote:

> Methods are excluded specifically to ensure generic classes and interfaces (such as Array<T>) continue to mostly relate covariantly. The impact of strictly checking methods would be a much bigger breaking change as a large number of generic types would become invariant
 **(even so, we may continue to explore this stricter mode).**
>

The above example has nothing to do with generic classes and an error that can easily be made especially with --strictNullChecks. Is it not possible to apply strictlyFunctionTypes checking to simple subclassing and not to generic. This is not a perfect solution, but better than now.

**Comment by chharvey** [comment]
For anyone new to this convo, I put together a [playground](https://www.typescriptlang.org/play/index.html#src=function%20method(run%3A%20(param%3A%20number%7Cstring)%20%3D%3E%20boolean)%3A%20void%20%7B%20%7D%0D%0A%0D%0Amethod(function%20(param%3A%20number)%20%7B%20return%20true%20%7D)%0D%0A%0D%0A%0D%0Aclass%20Parent%20%7B%20%0D%0A%20%20%20%20run(param%3A%20number%20%7C%20string)%3A%20boolean%20%7B%20return%20true%20%7D%0D%0A%7D%0D%0A%0D%0Aclass%20Child%20extends%20Parent%20%7B%0D%0A%20%20%20%20run(param%3A%20number)%3A%20boolean%20%7B%20return%20true%20%7D%0D%0A%7D%0D%0A) demonstrating this example.

With `--strictFunctionTypes` enabled, a function with parameter type `<S extends T>` is not assignable to a function with parameter type `<T>`, as expected:

```ts
function method(callback: (param: number|string) => boolean): void { }

method(function (param: number) { return true }) // error as expected
```

But method overloading with the same pattern does not give an error.

```ts
class Parent { 
    run(param: number | string): boolean { return true }
}

class Child extends Parent {
    run(param: number): boolean { return true } // no error!
}
```

Can’t we be consistent?
